File: image/src/main/kotlin/com/can/image/glide/GlideImageLoader.java
package com.can.image.glide;

import android.content.Context;
import android.support.annotation.DrawableRes;
import android.view.View;

import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.can.image.inter.OnProgressListener;
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView;

import java.io.File;
import java.lang.ref.WeakReference;

/**
 * Created by CAN on 19-1-4.
 *
 */

public class GlideImageLoader {

    private String url;
    private WeakReference<View> imageViewWeakReference;
    private GlideRequest<File> glideRequest;

    static GlideImageLoader create(View imageView) {
        return new GlideImageLoader(imageView);
    }

    private GlideImageLoader(View imageView) {
        imageViewWeakReference = new WeakReference<>(imageView);
        glideRequest = GlideApp.with(getContext()).asFile();
    }

    private View getImageView() {
        if (imageViewWeakReference != null) {
            return imageViewWeakReference.get();
        }
        return null;
    }

    public Context getContext() {
        if (getImageView() != null) {
            return getImageView().getContext();
        }
        return null;
    }

    public GlideRequest<File> getGlideRequest() {
        if (glideRequest == null) {
            glideRequest = GlideApp.with(getContext()).asFile();
        }
        return glideRequest;
    }


    private GlideRequest<File> loadImage(Object obj) {
        if (obj instanceof String) {
            url = (String) obj;
        }
        return glideRequest.load(obj);
    }

    public GlideImageLoader loadImage(Object obj, @DrawableRes int placeholder,OnProgressListener onProgressListener) {
        glideRequest = loadImage(obj);
        if (placeholder != 0) {
            glideRequest = glideRequest.placeholder(placeholder);
        }
        glideRequest.into(new GlideImageViewTarget<>(url, (SubsamplingScaleImageView) getImageView(),onProgressListener));
        return this;
    }


}

File: app/src/main/java/com/can/image/loadingeffect/glide/GlideImageLoader.kt
package com.can.image.loadingeffect.glide

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.net.Uri
import android.support.annotation.DrawableRes
import com.bumptech.glide.load.Transformation
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.transition.Transition
import com.can.image.`interface`.OnProgressListener
import com.can.image.glide.GlideApp
import com.can.image.glide.GlideRequest
import com.can.image.http.ProgressManager
import com.davemorrissey.labs.subscaleview.ImageSource
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
import java.io.File
import java.lang.ref.WeakReference

/**
 * Created by CAN on 19-1-5.
 *
 */
class GlideImageLoader<T> constructor(imageView: SubsamplingScaleImageView) {


    private var url: String? = null
    private var imageViewWeakReference: WeakReference<SubsamplingScaleImageView>? = null
    private var glideRequest: GlideRequest<T>? = null

    init {
        imageViewWeakReference = WeakReference(imageView)
    }

    fun getGlideRequest(): GlideRequest<T> {
        if (glideRequest == null) {
            glideRequest = GlideApp.with(getView()!!.context).asFile() as GlideRequest<T>
        }
        return glideRequest!!
    }

    private fun getView(): SubsamplingScaleImageView? {
        return if (imageViewWeakReference != null) {
            imageViewWeakReference!!.get()
        } else null
    }

    fun getUrl(): String {
        return if (url.isNullOrEmpty()) "" else url.toString()
    }

    fun loadImage(obj: Any, @DrawableRes placeholder: Int,onProgressListener: OnProgressListener?): GlideImageLoader<T> {
        if (placeholder != 0) {
            glideRequest = getGlideRequest().placeholder(placeholder)
        }

        if(onProgressListener!=null){
            if (obj is String) {
                url = obj
                ProgressManager().addListener(url as String, onProgressListener)
            }
        }

        if (getView() != null)
            getGlideRequest().load(obj).into(GlideImageViewTarget(getView()!!))
        return this
    }

    private inner class GlideImageViewTarget internal constructor(private val view: SubsamplingScaleImageView) : SimpleTarget<T>() {

        override fun onLoadFailed(errorDrawable: Drawable?) {
            val onProgressListener = ProgressManager().getProgressListener(getUrl())
            if (onProgressListener != null) {
                onProgressListener.onProgress(true, 100, 0, 0)
                ProgressManager().removeListener(getUrl())
            }
            super.onLoadFailed(errorDrawable)
        }

        override fun onResourceReady(file: T, transition: Transition<in T>?) {
            val onProgressListener = ProgressManager().getProgressListener(getUrl())
            if (onProgressListener != null) {
                onProgressListener.onProgress(true, 100, 0, 0)
                ProgressManager().removeListener(getUrl())
            }
            if(file is File){
                view.setImage(ImageSource.uri(Uri.fromFile(file as File)))
            }
        }
    }
}